### ISSUE TITLE
About batch and frame

### ISSUE THREAD

**Issue by cherry-ing** [j96w/DenseFusion#16]
Hi  @j96w ! Thank you for your work.
line 131~154 in train.py
```
            for i, data in enumerate(dataloader, 0):
                #…………
                train_count += 1

                if train_count % opt.batch_size == 0:
                    logger.info('Train time {0} Epoch {1} Batch {2} Frame {3} Avg_dis:{4}'.format(time.strftime("%Hh %Mm %Ss", time.gmtime(time.time() - st_time)), epoch, int(train_count / opt.batch_size), train_count, train_dis_avg / opt.batch_size))
```
I think one iteration is one batch, so batch number should be equivalent to `train_count`.
In `logger.info`, why batch number equals `int(train_count / opt.batch_size)`, and frame number equals `train_count`?

**Comment by j96w** [comment]
During training, one network update will be applied after `opt.batch_size` times loss backward. The `train_count` here refers to the times of loss backward. So the batch_num is `int(train_count / opt.batch_size)`.

**Comment by cherry-ing** [comment]
> During training, one network update will be applied after `opt.batch_size` times loss backward. The `train_count` here refers to the times of loss backward. So the batch_num is `int(train_count / opt.batch_size)`.

@j96w Thank you for your reply. Now I see, you set parameter `batch_size` to 1 for dataloader, which is not equivalent to `opt.batch_size` . Can I ask you why you did this? I mean what is the reason for calling `opt.batch_size` times `loss.backward()` before `optimizer.step()`,  instead of setting dataloader parameter `batch_size` to `opt.batch_size` and calling once `loss.backward()` in one batch.

**Comment by j96w** [comment]
To realize the pixel-wise dense fusion idea we introduced in this work, we need to keep the correspondence relationship between each RGB pixel and its corresponding depth (see the correspondence index ``choose`` return by the dataloader in each iteration). The main problem here about batch training is the size of the input RGB-D crops are different (with a large variance from 40x40 to 400x400 because of the different size of the objects). If we resize or use padding to make them the same size like most other works did, the correspondence relationship will be hard to maintain, since the resize of the depth image is different from the RGB resize and will lead to the deformation of the pointcloud generated from the resized depth. Thanks for mention this point, we have already got a buffering idea to improve the training efficiency of this code WRT this issue and will update soon.

**Comment by gaobaoding** [comment]
> To realize the pixel-wise dense fusion idea we introduced in this work, we need to keep the correspondence relationship between each RGB pixel and its corresponding depth (see the correspondence index `choose` return by the dataloader in each iteration). The main problem here about batch training is the size of the input RGB-D crops are different (with a large variance from 40x40 to 400x400 because of the different size of the objects). If we resize or use padding to make them the same size like most other works did, the correspondence relationship will be hard to maintain, since the resize of the depth image is different from the RGB resize and will lead to the deformation of the pointcloud generated from the resized depth. Thanks for mention this point, we have already got a buffering idea to improve the training efficiency of this code WRT this issue and will update soon.

Hi, @j96w ,
Thanks for your reply. But I am still confused of why you calling "opt.batch_size" times "loss.backward()" before "optimizer.step()", instead of "optimizer.step()" after each "loss.backward()" ?

**Comment by emigmo** [comment]
@gaobaoding hi, I can explain your confuse.
in pytorch, first "batch_size" times "loss.backward()"  and then follow "optimizer.step()", all the loss will be summed automatically and do one bp to update the parameters (SGD). but "optimizer.step()" after each "loss.backward()" would update the parameters at each time (GD). SGD will be better than GD, I think you know that.

**Comment by nanfeng-dada** [comment]
This is why I increased the batch_size, but the GPU memory usage rate did not change much, and the training speed did not increase much？

**Comment by zeal-github** [comment]
@j96w Hi, I have some question about resizing the image crop. If we do not consider the depth map. When we resize&crop, we are actually changing the camera intrinsic, right? But we have to maintain the identical intrinsic during training. So, does it possible to change the recorded gt pose according to the resize&crop arguments, so that we can get a 'visual pose' which is coincident with the crop.

**Comment by zx19950617** [comment]
> 
> 
> To realize the pixel-wise dense fusion idea we introduced in this work, we need to keep the correspondence relationship between each RGB pixel and its corresponding depth (see the correspondence index `choose` return by the dataloader in each iteration). The main problem here about batch training is the size of the input RGB-D crops are different (with a large variance from 40x40 to 400x400 because of the different size of the objects). If we resize or use padding to make them the same size like most other works did, the correspondence relationship will be hard to maintain, since the resize of the depth image is different from the RGB resize and will lead to the deformation of the pointcloud generated from the resized depth. Thanks for mention this point, we have already got a buffering idea to improve the training efficiency of this code WRT this issue and will update soon.

Hi, @j96w thanks for your excellent work and sharing code! 

The batch size in dataloader is always '1', which limits the training speed of the network, right? Now, have you any solution to increase the batch size in the dataloader to speed up the training process? About the buffering idea mentioned in your reply, I don't know how to achieve it. Have you updated the code for this idea? I'd like to refer to your solution for study.

Any suggestion will be appreciated a lot! Looking forward to your reply!
Best regards

**Comment by MengHao666** [comment]
> To realize the pixel-wise dense fusion idea we introduced in this work, we need to keep the correspondence relationship between each RGB pixel and its corresponding depth (see the correspondence index `choose` return by the dataloader in each iteration). The main problem here about batch training is the size of the input RGB-D crops are different (with a large variance from 40x40 to 400x400 because of the different size of the objects). If we resize or use padding to make them the same size like most other works did, the correspondence relationship will be hard to maintain, since the resize of the depth image is different from the RGB resize and will lead to the deformation of the pointcloud generated from the resized depth. Thanks for mention this point, we have already got a buffering idea to improve the training efficiency of this code WRT this issue and will update soon.

Have you released the version that take the **"buffering idea to improve the training efficiency"** idea ?

**Comment by opr-sergey** [comment]
> To realize the pixel-wise dense fusion idea we introduced in this work, we need to keep the correspondence relationship between each RGB pixel and its corresponding depth (see the correspondence index `choose` return by the dataloader in each iteration). The main problem here about batch training is the size of the input RGB-D crops are different (with a large variance from 40x40 to 400x400 because of the different size of the objects). If we resize or use padding to make them the same size like most other works did, the correspondence relationship will be hard to maintain, since the resize of the depth image is different from the RGB resize and will lead to the deformation of the pointcloud generated from the resized depth. Thanks for mention this point, we have already got a buffering idea to improve the training efficiency of this code WRT this issue and will update soon.

Hi @j96w, Thank you for sharing your project! 
Very interesting solution. How soon do you plan to update the code? thank you very much in advance
